comment/api/serializers.py

```diff
@@ -98,28 +99,52 @@ def get_reactions(obj):
 
 
 class CommentCreateSerializer(BaseCommentSerializer):
-
     class Meta:
         model = Comment
-        fields = ('id', 'user', 'content', 'parent', 'posted', 'edited', 'reply_count', 'replies', 'urlhash')
+        fields = ('id', 'user', 'email', 'content', 'parent', 'posted', 'edited', 'reply_count', 'replies', 'urlhash')
 
     def __init__(self, *args, **kwargs):
         self.model_type = kwargs['context'].get('model_type')
         self.model_id = kwargs['context'].get('model_id')
         self.user = kwargs['context'].get('user')
         self.parent_id = kwargs['context'].get('parent_id')
+        if self.context['request'].user.is_authenticated or (not settings.COMMENT_ALLOW_ANONYMOUS):
+            del self.fields['email']
+
         super().__init__(*args, **kwargs)
 
     def get_parent_object(self):
         if not self.parent_id or self.parent_id == '0':
             return None
         return Comment.objects.get(id=self.parent_id)
 
+    def validate_user(self, value):
+        request = self.context['request']
+        if request.user.is_authenticated:
+            return value
+
+    def validate_email(self, value):
+        request = self.context['request']
+        if request.user.is_authenticated:
+            value = request.user.email
+
+        elif settings.COMMENT_ALLOW_ANONYMOUS:
+            if (not value):
+                raise serializers.ValidationError(
+                    _('Email is required for posting anonymous comments'),
+                    code='invalid'
+                    )
+
+        return value.strip().lower()
+
     def create(self, validated_data):
+        # check if anonymous posting is allowed and user is unauthenticated
+        # send email by creating a key from the comment dict.
```

Comment:
Can we handle this the same way (I mean temporart `object` and `dict` creation) so that while confirming the comment, the key can be handled by the function `get_comment_from_dict` in `utils.py`?

I do understand that keys are different. I just wanted your take to know if something better is possible here.

Reply:
Yes. It's the best using this function `get_comment_from_key`.

Reply:
The keys for the dictionary required to form the link(key) is different here.

Reply:
I am not sure what you mean.

Reply:
In APIs, we have `model_id` and `model_name` from the HTTP requests while in the other one we use `model_name`, `model_app` and `app_name`. So, the method used for object(comment) creation is also different.

Reply:
I got confused, there is no model_app variable I think you meant model_id.
Cannot we get the app_name from model_type (this is the name of model_name in the api)?

Reply:
> Cannot we get the app_name from model_type (this is the name of model_name in the api)

i am not sure i completely understood what you meant.

Reply:
What I meant is that we can get the app name from the model name. However, I see that I am using 2 different ways to create a comment so it would be cleaner if we can use only one method to create a comment.

Reply:
can the same model name not be used in two different applications?

Reply:
Exactly. This will cause an issue. Therefore the app name must be provided in the api as well.

Reply:
It can be optional and we can perform the check if it's needed, then the delveloper will be asked to provide it

Reply:
I already opened this #46 to fix it.

Reply:
A one time check won't be sufficient. the check will have to be performed at every request. wouldn't it be better to have this information in the API request?

Reply:
okay, i will add this field in the API.

Reply:
> A one time check won't be sufficient. the check will have to be performed at every request. wouldn't it be better to have this information in the API request?

Yes. If the checks fails, we ask the user to provide the app name in url query string. However, you might make it required now and later we look for the optional choice.